### PR TITLE
Move security rules evaluation to StorageLayer

### PR DESCRIPTION
**Pull request by tonyjhuang**
### Description

Refactors security rules evaluation out of `firebase.ts` into `StorageLayer` for the following methods:

* Update object metadata
* Delete object
* List objects

### CODE REVIEW

File: src/emulator/storage/apis/gcloud.ts [hunk 1]
```diff
@@ -39,38 +40,54 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
...
+            decodedObjectId: req.params.objectId,
+          },
+          /* skipAuth = */ true
+        );
+      } catch (err) {
+        if (err instanceof NotFoundError) {
+          return res.sendStatus(404);
+        }
+        if (err instanceof ForbiddenError) {
+          throw new Error("Request failed unexpectedly due to Firebase Rules.");
```
Comment (tohhsinpei) [inline src/emulator/storage/apis/gcloud.ts:58]: Curious why we don't send a 403 here?

File: src/emulator/storage/files.ts [hunk 2]
```diff
@@ -177,11 +207,17 @@ export class StorageLayer {
...
+  public async handleGetObject(
+    request: GetObjectRequest,
+    skipAuth = false
+  ): Promise<GetObjectResponse> {
     const metadata = this.getMetadata(request.bucketId, request.decodedObjectId);
 
+    let authorized = skipAuth;
     // If a valid download token is present, skip Firebase Rules auth. Mainly used by the js sdk.
-    let authorized = (metadata?.downloadTokens || []).includes(request.downloadToken ?? "");
+    if (!authorized) {
```
Comment (tohhsinpei) [inline src/emulator/storage/files.ts:218]: Nit: I'd just do an or

File: src/emulator/storage/files.ts [hunk 3]
```diff
@@ -135,6 +135,36 @@ export type GetObjectResponse = {
...
+export type ListObjectsRequest = {
+  bucketId: string;
+  prefix: string;
+  delimiter: string;
+  pageToken?: string;
+  maxResults?: number;
+  authorization?: string;
+};
+
+/**  Response object for {@link StorageLayer#handleListObjects}. */
```
Comment (tohhsinpei) [inline src/emulator/storage/files.ts:163]: I probably wouldn't bother making a new type; just return a `ListResponse`

File: src/emulator/storage/apis/firebase.ts [hunk 4]
```diff
@@ -141,76 +152,57 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
...
-    if (
-      !(await isPermitted({
-        ruleset: emulator.rules,
-        method: RulesetOperationMethod.LIST,
-        path: operationPath,
-        file: {},
+    const maxResults = req.query.maxResults?.toString();
+    let response: ListObjectsResponse;
+    try {
+      response = await storageLayer.handleListObjects({
```
Comment (tohhsinpei) [inline src/emulator/storage/apis/firebase.ts:186]: Can this ever return a 404? What happens if I put a nonsense bucket id?

File: src/emulator/storage/apis/firebase.ts [hunk 5]
```diff
@@ -470,48 +464,35 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
   firebaseStorageAPI.put("/b/:bucketId/o/:objectId?", async (req, res) => {
     switch (req.header("x-http-method-override")?.toLowerCase()) {
       case "patch":
-        return handleMetadataUpdate(req, res);
+        return await handleMetadataUpdate(req, res);
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/emulator/storage/apis/firebase.ts:467]: good to know thanks!

File: src/emulator/storage/files.ts [hunk 6]
```diff
@@ -215,10 +136,10 @@ export class StorageLayer {
...
     let authorized = skipAuth;
     // If a valid download token is present, skip Firebase Rules auth. Mainly used by the js sdk.
-    if (!authorized) {
-      authorized = (metadata?.downloadTokens || []).includes(request.downloadToken ?? "");
-    }
-    if (!authorized) {
+    const hasValidDownloadToken = (metadata?.downloadTokens || []).includes(
+      request.downloadToken ?? ""
+    );
+    if (!authorized || hasValidDownloadToken) {
```
Comment (tohhsinpei) [inline src/emulator/storage/files.ts:142]: `!(authorized || hasValidDownloadToken)`

File: scripts/storage-emulator-integration/tests.ts [hunk 7]
```diff
@@ -1067,27 +1067,18 @@ describe("Storage emulator", () => {
...
-                  .on("end", () => {
-                    expect(Buffer.concat(data)).to.deep.equal(
-                      Buffer.from(IMAGE_FILE_BASE64, "base64")
-                    );
-                  })
-                  .on("close", resolve)
-                  .on("error", reject);
-              }
-            );
+            requestClient.get(downloadUrl, (response) => {
```
Comment (tohhsinpei) [inline scripts/storage-emulator-integration/tests.ts:1070]: Do you not need to pass an empty object as the options?

File: src/emulator/storage/apis/gcloud.ts [hunk 8]
```diff
@@ -292,6 +292,7 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
   });
 
   gcloudStorageAPI.get("/:bucketId/:objectId(**)", async (req, res) => {
+    console.log("this cant be right");
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/emulator/storage/apis/gcloud.ts:295]: done

File: scripts/storage-emulator-integration/tests.ts (hunk 7 above)
Comment (tonyjhuang) [reply scripts/storage-emulator-integration/tests.ts:1070]: no, `get` has two versions, a 3 arg version that takes options and 2 arg version that takes just the path and callback

File: src/emulator/storage/files.ts [hunk 9]
```diff
@@ -134,12 +134,12 @@ export class StorageLayer {
   ): Promise<GetObjectResponse> {
     const metadata = this.getMetadata(request.bucketId, request.decodedObjectId);
 
-    let authorized = skipAuth;
     // If a valid download token is present, skip Firebase Rules auth. Mainly used by the js sdk.
     const hasValidDownloadToken = (metadata?.downloadTokens || []).includes(
       request.downloadToken ?? ""
     );
-    if (!authorized || hasValidDownloadToken) {
+    let authorized = skipAuth || !!hasValidDownloadToken;
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/emulator/storage/files.ts]: done

File: src/emulator/storage/apis/firebase.ts [hunk 10]
```diff
@@ -104,7 +110,12 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
...
       } else if (err instanceof ForbiddenError) {
-        return res.sendStatus(403);
+        return res.status(403).json({
+          error: {
+            code: 403,
+            message: `Permission denied. No READ permission.`,
+          },
+        });
       }
       throw err;
```
Comment (abhis3) [inline src/emulator/storage/apis/firebase.ts:120]: Not entirely familiar of the standard for when we're throwing an error vs returning a status code, but could we throw a `400` error as the catch all case?

File: src/emulator/storage/apis/firebase.ts [hunk 11]
```diff
@@ -141,76 +152,57 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
...
-      !(await isPermitted({
-        ruleset: emulator.rules,
-        method: RulesetOperationMethod.UPDATE,
-        path: operationPath,
+    let metadata: StoredFileMetadata;
+    try {
+      metadata = await storageLayer.handleUpdateObjectMetadata({
+        bucketId: req.params.bucketId,
+        decodedObjectId: decodeURIComponent(req.params.objectId),
+        metadata: req.body as IncomingMetadata,
```
Comment (abhis3) [inline src/emulator/storage/apis/firebase.ts:160]: We should avoid type assertions

https://g3doc.corp.google.com/javascript/typescript/g3doc/dev/readability/styleguide.md?cl=head#type-and-non-nullability-assertions

File: src/emulator/storage/apis/gcloud.ts (hunk 1 above)
Comment (abhis3) [inline src/emulator/storage/apis/gcloud.ts:58]: Why are we not returning a `403` here?

File: src/emulator/storage/apis/firebase.ts (hunk 10 above)
Comment (tonyjhuang) [reply src/emulator/storage/apis/firebase.ts:120]: The intent of throwing an error here is to signal there is something fundamentally wrong with the emulator, ie an unexpected error case. Returning a 400 Bad Request can be confusing to users as they would assume something wrong with their request as opposed to something broken in the emulator. A better error code to return would be a 500 Internal, but raising an error and letting the emulator die is easier to debug than trying to understand why the "server" is failing

File: src/emulator/storage/apis/gcloud.ts (hunk 1 above)
Comment (tonyjhuang) [reply src/emulator/storage/apis/gcloud.ts:58]: Same comment as above, this is to signal a critical error in the emulator, as the current assumption is this endpoint should never fail with a 403

**Review by tohhsinpei — APPROVED** [review]

File: src/emulator/storage/apis/firebase.ts (hunk 10 above)
Comment (abhis3) [reply src/emulator/storage/apis/firebase.ts:120]: I see makes sense, thanks for the info!

**Pull request merged**